kinto/plugins/quotas/listener.py

```diff
@@ -0,0 +1,235 @@
+        obj_id = target['id']
+        parts = event_uri.split('/')
+        if resource_name in parts[-1]:
+            parts.append(obj_id)
+        else:
+            # Make sure the id is correct on grouped events.
+            parts[-1] = obj_id
+        uri = '/'.join(parts)
+
+        old = impacted.get('old', {})
+        new = impacted.get('new', {})
+
+        targets.append((uri, obj_id, old, new))
+
+    try:
+        bucket_info = copy.deepcopy(
+            storage.get("quota", bucket_uri, 'bucket_info'))
+    except RecordNotFoundError:
+        bucket_info = {
+            "collection_count": 0,
+            "record_count": 0,
+            "storage_size": 0,
+        }
+
+    collection_info = {
+        "record_count": 0,
+        "storage_size": 0,
+    }
+    if collection_id:
+        try:
+            collection_info = copy.deepcopy(
+                storage.get("quota", collection_uri, 'collection_info'))
+        except RecordNotFoundError:
+            pass
+
+    # Update the bucket quotas values for each impacted record.
+    for (uri, obj_id, old, new) in targets:
+        if resource_name == "collection":
+            collection_uri = uri
+        old_size = record_size(old)
+        new_size = record_size(new)
+
+        if max_bytes_per_item is not None and action != "delete":
+            if new_size > max_bytes_per_item:
+                message = ("MAX_BYTES_PER_ITEM size exceeded: %d Bytes. "
+                           "MAX_BYTES_PER_ITEM is %d Bytes" % (
+                               new_size, max_bytes_per_item))
+                raise http_error(HTTPInsufficientStorage(),
+                                 errno=ERRORS.FORBIDDEN.value,
+                                 message=message)
+
+        if action == 'create':
+            bucket_info['storage_size'] += new_size
+            if resource_name == 'collection':
+                bucket_info['collection_count'] += 1
+                collection_info['storage_size'] += new_size
+            if resource_name == 'record':
+                bucket_info['record_count'] += 1
+                collection_info['record_count'] += 1
+                collection_info['storage_size'] += new_size
+        elif action == 'update':
+            bucket_info['storage_size'] -= old_size
+            bucket_info['storage_size'] += new_size
+            if resource_name in ('collection', 'record'):
+                collection_info['storage_size'] -= old_size
+                collection_info['storage_size'] += new_size
+        elif action == 'delete':
+            bucket_info['storage_size'] -= old_size
+            if resource_name == 'collection':
+                bucket_info['collection_count'] -= 1
+                # When we delete the collection all the records in it
+                # are deleted without notification.
+                collection_records, _ = storage.get_all('record',
+                                                        collection_uri)
```

Comment:
* Add explicit argument names (`collection_id`, `parent_id`)
* How can you get the records if they were deleted then ?